internal/provider/resource_vcs_provider.go

```diff
@@ -0,0 +1,346 @@
+			OAuthClientID:     ptr.String(plan.OAuthClientID.ValueString()),
+			OAuthClientSecret: ptr.String(plan.OAuthClientSecret.ValueString()),
+		})
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error updating VCS provider",
+			err.Error(),
+		)
+		return
+	}
+
+	// Copy all fields returned by Tharsis back into the plan.
+	t.copyVCSProvider(*updated, &plan)
+
+	// Set the response state to the fully-populated plan, with or without error.
+	resp.Diagnostics.Append(resp.State.Set(ctx, plan)...)
+}
+
+func (t *vcsProviderResource) Delete(ctx context.Context,
+	req resource.DeleteRequest, resp *resource.DeleteResponse) {
+
+	// Get the current state.
+	var state VCSProviderModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Delete the VCS provider via Tharsis.
+	_, err := t.client.VCSProvider.DeleteProvider(ctx,
+		&ttypes.DeleteVCSProviderInput{
+			ID: state.ID.ValueString(),
+		})
+	if err != nil {
+
+		// Handle the case that the VCS provider no longer exists.
+		if tharsis.IsNotFoundError(err) {
+			resp.State.RemoveResource(ctx)
+			return
+		}
+
+		resp.Diagnostics.AddError(
+			"Error deleting VCS provider",
+			err.Error(),
+		)
+	}
+}
+
+// ImportState helps the provider implement the ResourceWithImportState interface.
+func (t *vcsProviderResource) ImportState(ctx context.Context,
+	req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
+
+	// Retrieve import ID and save to id attribute
+	resource.ImportStatePassthroughID(ctx, path.Root("id"), req, resp)
+}
+
+// copyVCSProvider copies the contents of a VCS provider.
+// It is intended to copy from a struct returned by Tharsis to a Terraform plan or state.
+func (t *vcsProviderResource) copyVCSProvider(src ttypes.VCSProvider, dest *VCSProviderModel) {
+	dest.ID = types.StringValue(src.Metadata.ID)
+	dest.Name = types.StringValue(src.Name)
+	dest.CreatedBy = types.StringValue(src.CreatedBy)
+	dest.Description = types.StringValue(src.Description)
+	dest.Hostname = types.StringValue(src.Hostname)
+	dest.GroupPath = types.StringValue(t.getParentPath(src.ResourcePath))
+	dest.ResourcePath = types.StringValue(src.ResourcePath)
+	dest.Type = types.StringValue(string(src.Type))
+	dest.AutoCreateWebhooks = types.BoolValue(src.AutoCreateWebhooks)
+	// The OAuthClientID and OAuthClientSecret fields are write-only to the Tharsis SDK, so no copying here.
+	// For the create operation, the OAuthAuthorizationURL field must be assigned by the caller.
+	// This just makes it not unknown, because Terraform requires computed fields to be known after apply.
+	dest.OAuthAuthorizationURL = types.StringValue("")
+
+	// Must use time value from SDK/API.  Using time.Now() is not reliable.
+	dest.LastUpdated = types.StringValue(src.Metadata.LastUpdatedTimestamp.Format(time.RFC850))
+}
+
+// getParentPath returns the parent path
+func (t *vcsProviderResource) getParentPath(fullPath string) string {
```

Comment:
This function isn't needed since the group path field is now exposed via the SDK (same for any other similar functions).